src/renderer/wg_engine/tvgWgShaderTypes.cpp

```diff
@@ -138,34 +138,41 @@ void WgShaderTypeGradient::update(const RadialGradient* radialGradient)
 void WgShaderTypeGradient::updateTexData(const Fill::ColorStop* stops, uint32_t stopCnt)
 {
     assert(stops);
-    // head
+    // sort input data by offset
+    static Array<Fill::ColorStop> sstops(stopCnt);
+    sstops.clear();
+    sstops.push(stops[0]);
+    // filter by increasing offset
+    for (uint32_t i = 1; i < stopCnt; i++)
+        if (sstops.last().offset < stops[i].offset)
+            sstops.push(stops[i]);
     uint32_t range_s = 0;
-    uint32_t range_e = uint32_t(stops[0].offset * (WG_TEXTURE_GRADIENT_SIZE-1));
+    uint32_t range_e = uint32_t(sstops[0].offset * (WG_TEXTURE_GRADIENT_SIZE-1));
     for (uint32_t ti = range_s; (ti < range_e) && (ti < WG_TEXTURE_GRADIENT_SIZE); ti++) {
-        texData[ti * 4 + 0] = stops[0].r;
-        texData[ti * 4 + 1] = stops[0].g;
-        texData[ti * 4 + 2] = stops[0].b;
-        texData[ti * 4 + 3] = stops[0].a;
+        texData[ti * 4 + 0] = sstops[0].r;
+        texData[ti * 4 + 1] = sstops[0].g;
+        texData[ti * 4 + 2] = sstops[0].b;
+        texData[ti * 4 + 3] = sstops[0].a;
     }
     // body
-    for (uint32_t di = 1; di < stopCnt; di++) {
-        range_s = uint32_t(stops[di-1].offset * (WG_TEXTURE_GRADIENT_SIZE-1));
-        range_e = uint32_t(stops[di-0].offset * (WG_TEXTURE_GRADIENT_SIZE-1));
+    for (uint32_t di = 1; di < sstops.count; di++) {
+        range_s = uint32_t(sstops[di-1].offset * (WG_TEXTURE_GRADIENT_SIZE-1));
+        range_e = uint32_t(sstops[di-0].offset * (WG_TEXTURE_GRADIENT_SIZE-1));
         for (uint32_t ti = range_s; (ti < range_e) && (ti < WG_TEXTURE_GRADIENT_SIZE); ti++) {
             float t = float(ti - range_s) / (range_e - range_s);
-            texData[ti * 4 + 0] = uint8_t((1.0f - t) * stops[di-1].r + t * stops[di].r);
-            texData[ti * 4 + 1] = uint8_t((1.0f - t) * stops[di-1].g + t * stops[di].g);
-            texData[ti * 4 + 2] = uint8_t((1.0f - t) * stops[di-1].b + t * stops[di].b);
-            texData[ti * 4 + 3] = uint8_t((1.0f - t) * stops[di-1].a + t * stops[di].a);
+            texData[ti * 4 + 0] = uint8_t((1.0f - t) * sstops[di-1].r + t * sstops[di].r);
+            texData[ti * 4 + 1] = uint8_t((1.0f - t) * sstops[di-1].g + t * sstops[di].g);
+            texData[ti * 4 + 2] = uint8_t((1.0f - t) * sstops[di-1].b + t * sstops[di].b);
+            texData[ti * 4 + 3] = uint8_t((1.0f - t) * sstops[di-1].a + t * sstops[di].a);
```

Comment:
don;t you want to use lerp?:
```
            texData[ti * 4 + 0] = lerp(sstops[di-1].r, sstops[di].r, t);
            texData[ti * 4 + 1] = lerp(sstops[di-1].g, sstops[di].g, t);
            texData[ti * 4 + 2] = lerp(sstops[di-1].b, sstops[di].b, t);
            texData[ti * 4 + 3] = lerp(sstops[di-1].a, sstops[di].a, t);
```


if not then maybe like that (less multiplications):
```
            texData[ti * 4 + 0] = (uint8_t)(sstops[di-1].r + (sstops[di].r - sstops[di-1].r) * t);
            texData[ti * 4 + 1] = (uint8_t)(sstops[di-1].g + (sstops[di].g - sstops[di-1].g) * t);
            texData[ti * 4 + 2] = (uint8_t)(sstops[di-1].b + (sstops[di].b - sstops[di-1].b) * t);
            texData[ti * 4 + 3] = (uint8_t)(sstops[di-1].a + (sstops[di].a - sstops[di-1].a) * t);
```